src/metabase/api/search.clj

```diff
@@ -374,23 +303,43 @@
             (if (number? v)
               (not (zero? v))
               v))]
-    (let [search-query {:union-all (for [model searchable-models
+    (let [search-query {:union-all (for [model search-config/searchable-models
                                          :let  [query (search-query-for-model model search-ctx)]
                                          :when (seq query)]
                                      query)}
           _            (log/tracef "Searching with query:\n%s" (u/pprint-to-str search-query))
-          results      (db/query search-query :max-rows search-max-results)]
-      ;; sort by [score model name]
-      (sort-by (juxt
-                (comp - (partial search/score (:search-string search-ctx)))
-                (comp model->sort-position :model)
-                :name)
-               (for [row results
-                     :when (check-permissions-for-model row)]
-                 ;; MySQL returns `:favorite` and `:archived` as `1` or `0` so convert those to boolean as needed
-                 (-> row
-                     (update :favorite bit->boolean)
-                     (update :archived bit->boolean)))))))
+          results      (db/reducible-query search-query :max-rows search-config/db-max-results)
+          cmp          (reify java.util.Comparator
+                         (compare [_ [score-1 _result-1] [score-2 _result-2]]
+                           (compare score-1 score-2)))
+          xf           (comp
+                        (filter check-permissions-for-model)
+                        ;; MySQL returns `:favorite` and `:archived` as `1` or `0` so convert those to boolean as needed
+                        (map #(update % :favorite bit->boolean))
+                        (map #(update % :archived bit->boolean))
+                        (map (partial scoring/score-and-result (:search-string search-ctx)))
+                        (filter some?))
+          rf           (fn queue-accumulator
```

Comment:
i'd probably put this accumulator into a utility namespace and have it take the size as a parameter. Then we could test it to show what its supposed to do and why its so good at it.

Reply:
also i suppose it will need to take the compare function (or default to `compare` i guess). And that's if you think its worth it to extract this. it does a lot of work inline here is why it kinda points out at me. so a docstring would probably be quite nice to give pointers.

Reply:
good call, just moved it (and `cmp`). Will write tests now